Step 3. Quality control
=======================

.. note:: 
    Consideration for the analysis.  
        Steps 3b and 3c can be done either in the interface or with the command-line code in the "merging.batch.harmony.R" in the project directory.
    Prerequisite
        Completed the step 1. foramtting and have the two files (matrix.csv and meta.data.csv files in the 1_SeuratQC folder) 

**STEP 3a.** Quality control of a Seurat object
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This section describes the Seurat QC process for human samples with 10x Genomics. I use the S133 sample from GES145370 dataset for the example screen shots. 

1. Upload both the meta-data.csv file and the matrix file (.csv.gz, .csv or .h5). 
2. Add the file name, which will be identical to the previous file name.
3. Select the data origin, as it will change the cut-off settings.
      - 10x Genomics >6000 features (transcripts per cell), are deemed to be doublets.
      - BD Rhapsody Immune panel > 150 features, are deemed to be doublets.
4.   Check that the file was uploaded
5.   Look at the first volcano plot "before" tab, as we need to check the cut-off for the MtDNA and rDNA cut-offs
      - MtDNA standard cut-off less than 20%
      - rRNA standard cut-off is greater than 5%. If there is no rRNA genes identifed, set value to 0 or the program will crash due to no cell's remaining.

.. image:: img/BeforeVolc.png
  :width: 600
  :alt: Alternative text

6. Once the settings have been configured, hit update. This will add the File name.
7. The user can then inspect the presence of the top variable features, elbow plot and Heatmaps. The latter two plots are to aid in setting the number of dimentions to use for the UMAP dimentional reduction. Default is set to 15. The resolutions are used for the unsupervised clustering. *Note: I use the scGate for the annotation stratergy rather than the unsupervised clusters as it is not suitable for identifying subtle differences in T cell sub-populations.*

.. image:: img/ElbowPlot.png
  :width: 600
  :alt: Alternative text

8. In the side panel, hit the "Run clustering" button to run the dimentionality reduction.
9. In the "Add meta data", first check that the file has uploaded correctly.
10. Next, In the side panel hit the "Input metatdata after clustering"
      - This will render a new table, and check that there are TCRs present. 
11. Now your file has completed the Seurat QC and has had the TCR sequences added to the meta-data. click the "Download Seurat" object and save it to the **"2_SCobj"** folder.
 
**STEP 3b.** Merging multiple Seurat objects 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This section is to merge multiple .rds objects and correct for the batch effect. 


1. From the **"2_SCobj"** upload all the .rds files.

.. image:: img/mergeSC.png
  :width: 600
  :alt: Alternative text

2. Download the merged.rds object

::

    require(STEGO.R)
    
    # large file merging and annotating
    # as the program window may disappear, we recommend merging and annotating using the following commands
    
    ###### merging seurat object ------
    # Check that you are in the correct working directory with your RDS files
    merging_multi_SeuratRDS(set_directory = "2_scObj/", merge_RDS = F, pattern_RDS = ".rds$") 
    
    # once that is check, switch merge_RDS to TRUE or T
    sc_merge <- merging_multi_SeuratRDS(set_directory = "2_scObj/", merge_RDS = T, pattern_RDS = ".rds$") 
    
    # merges the different layers in V5 of seurat 
    sc <- JoinLayers(sc_merge,  assay = "RNA") 
    
    # save the merged file - it will not have the scaled data and PCA stored any more due to the merging process. This is in case R crashes and it needs to be read in
    saveRDS(sc_merge,"2_scObj/sc_merge.rds") 

**STEP 3c.** Batch correction
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Same processes as Seurat and see https://satijalab.org/seurat/articles/pbmc3k_tutorial.html for details. 

2. Wait for the files to finish uploading before proceeding 
3. Update the Project name e.g., BreastCancer
4. Hit the bottons in each tab in the following order:
    - Run VariableFeatures
    - Run Scale
    - Run PCA
    - Run Harmony *Note: Harmony R package is used for the batch correction.* 
        * https://www.nature.com/articles/s41592-019-0619-0 for more detail
    - Run Dimentional Reduction

5. Once completed click the "Download merged Seurat"

*Note: Depending on the number of files, and total number of cells will impact the speed of merging*

If you have more than a few files to merge and annotate, we recommend using the following code: 

::

    require(STEGO.R)

    sc <- readRDS("2_scObj/sc_merge.rds")

    ## perform the harmony batch correction ------
    sc <- harmony_batch_correction_1_variableFeatures(file = sc)
    sc <- harmony_batch_correction_2_Scaling(file = sc, Seruat_version = "V5")
    sc <- harmony_batch_correction_3_PC(file = sc)
    sc <- harmony_batch_correction_4_Harmony(file = sc)
    saveRDS(sc,"2_scObj/sc_harmony.rds")


**Step 3d.** Annotating the Seurat object
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This section describes how to annotated the files either from the "default" and/or "custom" modules with scGate. If you have large files, we recommend using scripts and not the interface. 

::

    require(STEGO.R)

    #### annotating Seurat object -----
    sc <- readRDS("2_scObj/sc_harmony.rds")

    sc <-     scGate_annotating(
              file = sc,
              TcellFunction = T,
              generic = T,
              exhausted = T,
              senescence = T,
              cycling = T,
              Th1_cytokines = T,
              TCRseq = T,
              threshold = 0.2, # change to 0.2 if you use the focused immuen panel from BD rhapsody
              reductionType = "harmony"
            )
    
    sc@meta.data$Cell_Index_old <- sc@meta.data$Cell_Index
    sc@meta.data$Cell_Index <- rownames(sc@meta.data)
    head(sc@meta.data)
    saveRDS(sc,"3_analysis/sc_anno.rds")


Default annotation models with scGate
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

*Note: Under development.*

The scGate threshold was set to 0.2 for full capture and 0.5 for the immune panel (<400 transcripts). 

* **Human**
    - TcellFunction
    - Generic
    - exhausted
    - senescence
    - cycling
    - Th1_cytokines (IFNg and TNF)
    - TCRseq (meta.data TCR-seq)

* **10X Genomcis for Mouse** (*Will likely be the same as the BD Rhapsody mouse full panel model.*)

* **BD Rhapsody Human immune panel**

* **Mouse Panel** Under development

TCR annotations
^^^^^^^^^^^^^^^

This section annotates the potential invariant T cells

Functional T cell annotations
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


.. csv-table:: Annotation_table
    :header-rows: 1
    :file: path/anno_table.csv
    :widths: 20, 30, 50
    :class: longtable

Cyto = GZMA,GZMB,PRF1,GNLY,NKG7; Naive = CCR7, SELL

DN = double negative; Th = T helper, Tfh = follicular T helper, Tc = cytotoxic T cell; Treg = regulatory T cell 

Creating custom annotation stratergies
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

*Note: section is incomplete.*

There are nine custom databases that a user can amend. If needed, these can be added to an already annotated file.

**Design your custom marker set**

Check that the cells express the markers under the "marker check" tab.
    - To run, click the 'View Feature plot'
    - The user can alter the range to a custom 

If needed, you can use either the "Treatment differences within cluster" or "cluster difference" to have better identification of marker differneces between clusters.



